src/main.rs

```diff
@@ -273,12 +275,58 @@ fn main() {
 	impl BuildStatus {
 		fn set_status(&self, worker: usize, status: WorkerStatus) {
 			let mut lock = self.inner.lock().unwrap();
-			lock.workers[worker] = status;
-			lock.dirty = true;
+			lock.set_status(worker, status);
 			self.condvar.notify_all();
 		}
 	}
 
+	impl BuildStatusInner {
+		fn set_status(&mut self, worker: usize, status: WorkerStatus) {
+			self.workers[worker] = status;
+			self.dirty = true;
+		}
+
+		fn all_workers_done(&self) -> bool {
+			self.workers.iter().all(|worker| *worker == WorkerStatus::Done)
+		}
+
+		fn all_workers_done_or_idle(&self) -> bool {
+			self.workers.iter().all(|worker| *worker == WorkerStatus::Done || *worker == WorkerStatus::Idle)
+		}
+	}
+
+	// A lambda that returns the estimated total time for an average task, if possible.
+	let average_total_task_time_func = || -> Option<Duration> {
+		match build_log.average_historic_task_time() {
+			Some(duration) => {
+				Some(duration)
+			},
+			None => {
+				// TODO: Try to find some estimate using a running average on the tasks that ran until now?
+				None
+			},
+		}
+	};
+
+	// A lambda that returns the estimated total time for a given task, if possible.
+	let estimated_total_task_time_func = |task: usize| -> Option<Duration> {
+		let rule = &spec.build_rules[task];
+		let command = match &rule.command {
+			BuildRuleCommand::Phony => unreachable!("Got phony task."),
+			BuildRuleCommand::Command { command, .. } => command,
+		};
+
+		if rule.outputs.len() == 0 {
+			average_total_task_time_func()
+		} else {
+			let output = &rule.outputs[0];
+			match build_log.estimated_total_task_time(output.clone(), command.clone()) {
+				Some(duration) => Some(duration),
+				None => average_total_task_time_func(),
+			}
```

Comment:
This match is equivalent to `.unwrap_or_else(average_total_task_time_func)`: [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else)